src/plugins/extra/smartquotes.rs

```diff
@@ -0,0 +1,517 @@
+    /// If a match is found, returns `Some` with two `ReplacementOp`s to be
+    /// added to the result, and with the resulting length of the `quote_stack`.
+    fn try_close(
+        quote_stack: &Vec<QuoteMarker>,
+        walk_index: usize,
+        level: u32,
+        quote_type: QuoteType,
+        quote_position: usize,
+    ) -> Option<(ReplacementOp, ReplacementOp, usize)> {
+        for (j, other_item) in quote_stack.iter().enumerate().rev() {
+            if other_item.level < level {
+                return None;
+            }
+            if other_item.quote_type == quote_type && other_item.level == level {
+                return Some((
+                    ReplacementOp {
+                        walk_index: other_item.walk_index,
+                        quote_position: other_item.quote_position,
+                        quote: if quote_type == QuoteType::Single {
+                            OPEN_SINGLE_QUOTE
+                        } else {
+                            OPEN_DOUBLE_QUOTE
+                        },
+                    },
+                    ReplacementOp {
+                        walk_index,
+                        quote_position,
+                        quote: if quote_type == QuoteType::Single {
+                            CLOSE_SINGLE_QUOTE
+                        } else {
+                            CLOSE_DOUBLE_QUOTE
+                        },
+                    },
+                    j,
+                ));
+            }
+        }
+        None
+    }
+}
+
+/// Produces a simplified flat list of all tokens, with the necessary
+/// information to do smart quote replacement.
+///
+/// This handles inline html and inline code like JS version seems to do.
+/// This list is a work-around for the fact that we can't build a flat list of
+/// all nodes for iteration back and forth, and at the same time do a mutable
+/// walk on the document tree.
+///
+/// Returns a `Vec<FlatToken<'a>>` where `<'a>` is the same lifetime as `root`.
+/// This simply reflects the fact that the `content: &str` entries of the
+/// `FlatToken` structs reference the same memory as `root`'s children.
+/// Every entry in the `Vec` will produce an entry in the result, meaning that
+/// the index of a token in the resulting `Vec` will be the same as the index it
+/// would get during a `root.walk` call.
+fn all_text_tokens<'a>(root: &'a Node) -> Vec<FlatToken<'a>> {
+    let mut result = Vec::new();
+    let mut walk_index = 0;
+    root.walk(|node, nesting_level| {
+        if let Some(text_node) = node.cast::<Text>() {
+            result.push(FlatToken::Text {
+                content: &text_node.content,
+                nesting_level,
+            });
+        } else if let Some(html_node) = node.cast::<HtmlInline>() {
+            result.push(FlatToken::Text {
+                content: &html_node.content,
+                nesting_level,
+            });
+        } else if let Some(_) = node.cast::<Paragraph>() {
```

Comment:
I made `node.is::<Paragraph>()` function specifically for this reason.

it probably looks better than `node.cast::<Paragraph>().is_some()` like clippy would suggest